File: Services/pkg/user_role/service.go
package user_role

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/domain"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	helpers "github.com/zercle/gofiber-helpers"
)

type userRoleService struct {
	repository domain.UserRoleRepository
}

func NewUserRoleService(repository domain.UserRoleRepository) domain.UserRoleService {
	return &userRoleService{repository: repository}
}

func (s *userRoleService) CreateUserRole(ctx context.Context, userRole models.UserRole) []helpers.ResponseError {
	err := s.repository.CreateUserRole(ctx, userRole)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

func (s *userRoleService) GetUserRole(ctx context.Context, id uint) (*models.UserRole, []helpers.ResponseError) {
	userRole, err := s.repository.GetUserRole(ctx, id)
	if err != nil {
		return nil, []helpers.ResponseError{*err}
	}
	return userRole, nil
}

func (s *userRoleService) GetUserRoles(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.UserRole, *models.Pagination, *models.Search, []helpers.ResponseError) {
	userRoles, paginated, searched, err := s.repository.GetUserRoles(ctx, pagination, search)
	if err != nil {
		return nil, nil, nil, []helpers.ResponseError{*err}
	}
	return userRoles, paginated, searched, nil
}

func (s *userRoleService) GetUserRolesByUserID(ctx context.Context, userID uint) ([]models.UserRole, []helpers.ResponseError) {
	userRoles, err := s.repository.GetUserRolesByUserID(ctx, userID)
	if err != nil {
		return nil, []helpers.ResponseError{*err}
	}
	return userRoles, nil
}

func (s *userRoleService) UpdateUserRole(ctx context.Context, id uint, userRole models.UserRole) []helpers.ResponseError {
	err := s.repository.UpdateUserRole(ctx, id, userRole)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

func (s *userRoleService) DeleteUserRole(ctx context.Context, id uint) []helpers.ResponseError {
	err := s.repository.DeleteUserRole(ctx, id)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

func (s *userRoleService) DeleteUserRolesByUserID(ctx context.Context, userID uint) []helpers.ResponseError {
	err := s.repository.DeleteUserRolesByUserID(ctx, userID)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

File: hono-clean-architecture/pkg/user_role/service.ts
import { UserRole, Pagination, Search } from "../models";
import { UserRoleRepository, UserRoleService as IUserRoleService } from "../domain";
import { ResponseError } from "../../internal/infrastructure/custom_error";

export class UserRoleService implements IUserRoleService {
  private repository: UserRoleRepository;

  constructor(repository: UserRoleRepository) {
    this.repository = repository;
  }

  async createUserRole(data: { userId: number; roleId: number }): Promise<ResponseError[]> {
    const error = await this.repository.createUserRole(data);
    if (error) {
      return [error];
    }
    return [];
  }

  async getUserRole(id: number): Promise<{ userRole: UserRole | null; errors: ResponseError[] }> {
    const { userRole, error } = await this.repository.getUserRole(id);
    if (error) {
      return { userRole: null, errors: [error] };
    }
    return { userRole, errors: [] };
  }

  async getUserRoles(
    pagination: Pagination,
    search: Search
  ): Promise<{
    userRoles: UserRole[];
    pagination: Pagination;
    search: Search;
    errors: ResponseError[];
  }> {
    const result = await this.repository.getUserRoles(pagination, search);
    if (result.error) {
      return { userRoles: [], pagination, search, errors: [result.error] };
    }
    return {
      userRoles: result.userRoles,
      pagination: result.pagination,
      search: result.search,
      errors: [],
    };
  }

  async deleteUserRole(id: number): Promise<ResponseError[]> {
    const error = await this.repository.deleteUserRole(id);
    if (error) {
      return [error];
    }
    return [];
  }
}

export function newUserRoleService(repository: UserRoleRepository): IUserRoleService {
  return new UserRoleService(repository);
}
